### PR TITLE
Improve handling when focused button becomes disabled

### LINKED ISSUE #1935: FocusRing sticks around on buttons
# 🐛 Bug Report 

https://codesandbox.io/s/serverless-haze-3i3ng?file=/src/App.js
Edit: I've update the codesandbox to handle the issue how we'd expect it to be handled

Tab to first button
Hit Enter
Tab to second button
First button retains focusring

The focusring should go away when disabled

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

### PR DESCRIPTION
**Pull request by devongovett**
Closes #1935

This fixes two issues:

1. Focus ring not disappearing when a focused button becomes disabled. That is due to a combination of a React issue, where onBlur is not fired if a button becomes disabled, and the same issue natively in Firefox. This is solved by using native blur events, as well as a `MutationObserver` to simulate those for Firefox.
2. In calendar, when the previous and next buttons become disabled as a result of pressing them, focus is lost. Now we send it to the calendar body instead.

### CODE REVIEW

File: packages/@react-aria/interactions/src/useFocusWithin.ts [hunk 1]
```diff
@@ -56,30 +57,74 @@ export function useFocusWithin(props: FocusWithinProps): FocusWithinResult {
...
+          observer = null;
+        }
+      };
+
+      e.currentTarget.addEventListener('focusout', onBlur);
+
+      // If this is a <button> or <input>, use a MutationObserver to watch for the disabled attribute.
+      // This is necessary because Firefox does not fire blur/focusout in this case. We dispatch these events ourselves instead.
+      // For browsers that do, focusout fires before the MutationObserver, so onBlur should not fire twice.
+      if (e.target instanceof HTMLButtonElement || e.target instanceof HTMLInputElement) {
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/useFocusWithin.ts:96]: HTMLTextareaElement as well i think
HTMLFormElement too? (for landmarks eventually)

I think it's all form fields if I recall the description correctly. But I think those four are the only ones we actually deal with

File: packages/@react-aria/interactions/src/useFocusWithin.ts [hunk 2]
```diff
@@ -56,30 +57,74 @@ export function useFocusWithin(props: FocusWithinProps): FocusWithinResult {
...
+        let target = e.target;
+        observer = new MutationObserver(() => {
+          if (state.isFocusWithin && target.disabled) {
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/useFocusWithin.ts:99]: what happens if the element is unmounted while focus is on it? It looks like the mutation observer will be left around? not sure if it fires when the element it's observing is about to be removed

File: packages/@react-aria/interactions/src/utils.ts [hunk 3]
```diff
@@ -29,3 +32,114 @@ export function isVirtualClick(event: MouseEvent | PointerEvent): boolean {
...
+  let stateRef = useRef({
+    isFocused: false,
+    onBlur,
+    observer: null as MutationObserver
+  });
+  let state = stateRef.current;
+  state.onBlur = onBlur;
+
+  // Clean up MutationObserver on unmount. See below.
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/interactions/src/utils.ts:95]: Oops, yeah thanks.

File: packages/@react-aria/interactions/src/utils.ts [hunk 4]
```diff
@@ -29,3 +32,117 @@ export function isVirtualClick(event: MouseEvent | PointerEvent): boolean {
...
+  let stateRef = useRef({
+    isFocused: false,
+    onBlur,
+    observer: null as MutationObserver
+  });
+  let state = stateRef.current;
+  state.onBlur = onBlur;
+
+  // Clean up MutationObserver on unmount. See below.
+  // eslint-disable-next-line arrow-body-style
```
Comment (devongovett) [inline packages/@react-aria/interactions/src/utils.ts:94]: This is a silly lint rule and we should consider removing it.

Comment (snowystinger) [reply packages/@react-aria/interactions/src/utils.ts:94]: gotta wait until we find a new lint rule to add =P

**Review by snowystinger — DISMISSED** [review]
Tested using https://reactspectrum.blob.core.windows.net/reactspectrum/7db24be038548e83d89e7bae38dbcce04206e162/storybook/index.html?path=/story/date-and-time-calendar--minvalue-today-visiblemonths-3
Works in Chrome/Safari/FF with keyboard navigation

**Comment by devongovett** [comment]
Good catch @LFDanLu! Looks like React is not batching those updates somehow, so the order matters.

**Review by snowystinger — APPROVED** [review]
Good catch @LFDanLu

**Review by LFDanLu — APPROVED** [review]
LGTM, things seems to behave nicely now

**Pull request merged**